### ISSUE TITLE
Global mode doesn't work via update override or pre-deploy script

### ISSUE THREAD

**Issue by VoidMonk** [caprover/caprover#1699]
**IMPORTANT:** This is a bug report. If you are having problem with deploying a particular app use the deployment issue type.

> If you are having problems with deploying a particular app using a customized dockerfile - please change your questions to Dockerfile and post it where appropriate such as StackOverflow. This is a Docker question, not a CapRover specific issue.

---

What is the problem?

Service Update Override and Pre-Deploy Script don't allow swarm global mode instead of the default replicated mode.

If applicable, content of `captain-definition` file:

NA

Steps to reproduce the problem:

Specify yaml in App Config's 'Service Update Override', as per [docs](https://caprover.com/docs/service-update-override.html) and [Docker API spec](https://docs.docker.com/engine/api/v1.40/#tag/Service/operation/ServiceUpdate), to enable global mode for a service:

```yaml
Mode:
  Global: {}
```

or JSON:

```json
{
  "Mode": {
    "Global": {}
  }
}
```

The UI shows an error: `Something bad happened.`

Even specifying a pre-deploy script, as per [docs](https://caprover.com/docs/pre-deploy-script.html), gives the same error message:

```js
var preDeployFunction = function (captainAppObj, dockerUpdateObject) {
    return Promise.resolve()
        .then(function() {
            //dockerUpdateObject.TaskTemplate.ForceUpdate = 1;
            dockerUpdateObject.Mode.Global = {};
            dockerUpdateObject.Mode.Replicated = null;
            return dockerUpdateObject;
        });
};
```

It'll be even better if App Configs page had a toggle between global and replicated modes (with instance/replica count).

Output of the following command on your server:

```
uname -a && lsb_release -a && free -h
```

OS: Linux (Ubuntu 20.04 LTS)
Architecture: x86_64
Docker Version: 23.0.1

**Comment by githubsaturn** [comment]
You can always get more logs from the server to see what went wrong: `docker service logs captain-captain --follow`

In this case, here is what happened:
If you use the service override, the previous field (Replicated) doesn't get set to null, and you can't use both Replicated and Global at the same time


If you run the predeploy script (and correctly setting Replicated object to null), you'll see this error
```
service mode change is not allowed
```

And here is the cause:
- https://stackoverflow.com/questions/51297393/docker-swarm-service-change-the-service-mode-from-replicas-to-global-without-r

In short, this feature is not supported by Docker

**Comment by VoidMonk** [comment]
Thanks for clarifying.

Strangely, the Docker API spec for [service update](https://docs.docker.com/engine/api/v1.40/#tag/Service/operation/ServiceUpdate) request body schema shows that `Mode` can be set to `Global` (object), which is confusing as to why they would include it if it's not supported, or perhaps it was added more recently that the last CapRover review and update. I'll really appreciate if you can double-check, for possible support in CapRover.

I also came across this article: `[Manage Docker Swarm with Portainer](https://thenewstack.io/tutorial-manage-docker-swarm-with-portainer/)`, and noticed in one of the screenshots (figure 2 on the webpage) that Portainer seems to support both modes, at-least when creating a service, which will be a good feature to have in CapRover too.

Without global mode support, there's really no way to have exactly one instance of the service on all nodes.

**Comment by githubsaturn** [comment]
Yea unfortunately this is a shortcoming of docker. You can try this hack to see if it helps:
- create a glovlval service using docker CLI. Call it srv-captain--myapp
- create an app called myapp
- make the required changes in caprover predeploy script

It might work.

**Comment by VoidMonk** [comment]
Thanks for the suggested workaround.

I created a global service using Docker CLI, then created an app in CapRover, but updating the app from the UI with pre-deploy script or instance count (e.g. 0, -1 etc.) etc. throws an error as per server logs: `Error: Non replicated services cannot be associated with instance count`.

This is due to an 'instance count' validation in DockerApi.ts (https://github.com/caprover/caprover/blob/master/src/docker/DockerApi.ts#L1478).

Any way to get around this, or fix the validation in CapRover itself?

**Comment by githubsaturn** [comment]
Looks like the check will fail if instanceCount is -1. Try putting -1 for instanceCount.

**Comment by VoidMonk** [comment]
That's what I thought looking at the code, but even setting the instance count in the UI to -1 or any other negative number doesn't work. Same error message in docker logs.

Seems like an input sanitization elsewhere is reverting the negative instance count to a positive one, causing that validation to fail.

Service update overrides and pre-deploy script take affect after the instance count validation, so they won't work unless this is somehow bypassed for global mode services, which is a pretty common swarm scenario.

**Comment by VoidMonk** [comment]
If I were to submit a PR with a simple non-disruptive validation fix (please see my fork commit at https://github.com/VoidMonk/caprover/commit/147803261ba1e1a1541799adfd982b077c3ccf5f), then is it possible for it to be approved and pushed to the `edge` channel sooner than later?

**Comment by githubsaturn** [comment]
Looks good. Please submit!
